### PR TITLE
Remove `feed_local_def_id`

### PR DESCRIPTION
**Pull request by oli-obk**
best reviewed commit by commit

Basically I returned `TyCtxtFeed` from `create_def` and then preserved that in the local caches

based on https://github.com/rust-lang/rust/pull/121084

r? @petrochenkov

### CODE REVIEW

**Comment by oli-obk** [comment]
I think [56ef997](https://github.com/rust-lang/rust/pull/121089/commits/56ef997a7a35915fba74495a8dc8c2400abfd6c8) is unnecessary, need to check

**Comment by petrochenkov** [comment]
I don't understand what we get from this, to be honest.
What are the additional guarantees from keeping `Feed`s in `node_id_to_def_id` tables?

**Comment by oli-obk** [comment]
> What are the additional guarantees from keeping `Feed`s in `node_id_to_def_id` tables?

that you don't feed from another query but the one that created the `DefId`. `Feed` cannot be transferred across query boundaries

**Comment by petrochenkov** [comment]
> > What are the additional guarantees from keeping `Feed`s in `node_id_to_def_id` tables?
> 
> that you don't feed from another query but the one that created the `DefId`. `Feed` cannot be transferred across query boundaries.
> 
> `feed_local_def_id` is a massive footgun and if anone uses it from anywhere but the resolver it will cause undebuggable incremental ICEs. The reason the resolver is fine is that it is (like the `analysis` query, an outermost (not called from another query) `eval_always` query.

I probably makes sense to put this explanation as a comment on `struct Feed`.

File: compiler/rustc_middle/src/ty/context.rs [hunk 1]
```diff
@@ -522,15 +523,33 @@ pub struct TyCtxtFeed<'tcx, KEY: Copy> {
...
     pub fn feed_unit_query(self) -> TyCtxtFeed<'tcx, ()> {
         TyCtxtFeed { tcx: self, key: () }
     }
     pub fn feed_local_crate(self) -> TyCtxtFeed<'tcx, CrateNum> {
         TyCtxtFeed { tcx: self, key: LOCAL_CRATE }
     }
-    pub fn feed_local_def_id(self, key: LocalDefId) -> TyCtxtFeed<'tcx, LocalDefId> {
-        TyCtxtFeed { tcx: self, key }
+
+    pub fn feed_local_crate_def_id(self) -> TyCtxtFeed<'tcx, LocalDefId> {
```
Comment (petrochenkov) [inline compiler/rustc_middle/src/ty/context.rs:551]: Why is `CRATE_DEF_ID` different from other def ids?
It is also created by one query (`resolver_for_lowering`) and supposedly cannot be used inside any other query.

**Comment by petrochenkov** [comment]
>`Feed` cannot be transferred across query boundaries.

I'm not sure if it's a hard guarantee, or something that we need to keep in mind when writing new code.

For example, what happens if we put `Feed` into `ResolverOutputs`?
Then it will be usable from other queries through `tcx.resolutions(())`.

**Comment by oli-obk** [comment]
> I'm not sure if it's a hard guarantee

I'll add a negative impl for the relevant traits

File: compiler/rustc_middle/src/ty/context.rs (hunk 1 above)
Comment (oli-obk) [reply compiler/rustc_middle/src/ty/context.rs:551]: it's created before the `TyCtxt` is created: https://github.com/rust-lang/rust/blob/master/compiler/rustc_interface/src/queries.rs#L149

We could probably move everything about CRATE_DEF_ID into `resolver_for_lowering` and actually create it there (which would give us a `Feed` to then feed queries from there)

Comment (oli-obk) [reply compiler/rustc_middle/src/ty/context.rs]: But you're right in that this is not nice and clean, so I refactored it to remove all but the pre-query feeding uses

**Comment by oli-obk** [comment]
@rust-timer build 9b35afffcbf0909e8881fbf013bc6f921fae936c

**Comment by oli-obk** [comment]
@rust-timer build 9b35afffcbf0909e8881fbf013bc6f921fae936c

**Comment by oli-obk** [comment]
No perf changes: https://perf.rust-lang.org/compare.html?start=e29a1530f670d66f617f7aac8601920a87263ac6&end=9b35afffcbf0909e8881fbf013bc6f921fae936c

**Comment by oli-obk** [comment]
This is ready for review again

File: compiler/rustc_interface/src/passes.rs [hunk 2]
```diff
@@ -608,6 +617,8 @@ pub static DEFAULT_QUERY_PROVIDERS: LazyLock<Providers> = LazyLock::new(|| {
     providers.analysis = analysis;
     providers.hir_crate = rustc_ast_lowering::lower_to_hir;
     providers.resolver_for_lowering = resolver_for_lowering;
+    providers.stripped_cfg_items = stripped_cfg_items;
+    providers.resolutions = resolutions;
```
Comment (petrochenkov) [inline compiler/rustc_interface/src/passes.rs:621]: `stripped_cfg_items` and `resolutions` are trivial and used once, they can be turned into closures here.

File: compiler/rustc_middle/src/ty/mod.rs [hunk 3]
```diff
@@ -28,6 +28,7 @@ use crate::ty::fast_reject::SimplifiedType;
 use crate::ty::util::Discr;
 pub use adt::*;
 pub use assoc::*;
+use ast::expand::StrippedCfgItem;
```
Comment (petrochenkov) [inline compiler/rustc_middle/src/ty/mod.rs:31]:
```suggestion
use rustc_ast::expand::StrippedCfgItem;
```
Editor auto importing strikes again.

File: compiler/rustc_resolve/src/def_collector.rs [hunk 4]
```diff
@@ -26,26 +27,38 @@ struct DefCollector<'a, 'b, 'tcx> {
 }
 
 impl<'a, 'b, 'tcx> DefCollector<'a, 'b, 'tcx> {
-    fn create_def(
+    fn create_def_with_feed(
```
Comment (petrochenkov) [inline compiler/rustc_resolve/src/def_collector.rs:30]: This function is used once and can be inlined and removed.

**Comment by petrochenkov** [comment]
Left some style comments.
@rustbot author

**Comment by petrochenkov** [comment]
Are there any plans to support feeding queries from other queries later?
The current restrictions are probably going to cause problems once earlier compilation is querified, e.g. late name resolution.

**Comment by oli-obk** [comment]
> Are there any plans to support feeding queries from other queries later?

No. That is not possible to support correctly in general. There are situations where it can be fine, and once we have them, we can look into supporting them properly.

But: once something is querified better, it often becomes possible to just return the values you need from the appropriate queries and make the queries you want to feed invoke the queries that have the values instead.

**Comment by petrochenkov** [comment]
r=me after updating the failing test.
@rustbot author

File: compiler/rustc_resolve/src/lib.rs [hunk 5]
```diff
@@ -1332,7 +1340,8 @@ impl<'a, 'tcx> Resolver<'a, 'tcx> {
         let mut def_id_to_node_id = IndexVec::default();
         assert_eq!(def_id_to_node_id.push(CRATE_NODE_ID), CRATE_DEF_ID);
         let mut node_id_to_def_id = NodeMap::default();
-        node_id_to_def_id.insert(CRATE_NODE_ID, CRATE_DEF_ID);
+        let crate_feed = tcx.feed_local_def_id(CRATE_DEF_ID).downgrade();
```
Comment (spastorino) [inline compiler/rustc_resolve/src/lib.rs]: A little bit late but this should probably be crate_def_id

**Pull request merged**